src/combineReducers.js

```diff
@@ -1,9 +1,66 @@
 import Immutable from 'seamless-immutable';
+import {
+  assertReducerShape,
+  getUndefinedStateError,
+  getNoValidReducersError,
+  getNotSupportedTypeAsReducerError,
+  getPossibleUnexpectedStateShapeWarning
+} from './utils/combineReducersValidation'
 
 export default function combineReducers(reducers) {
-  return (state=Immutable({}), action) => Object.keys(reducers)
-    .reduce(
-      (iState, key) => iState.set(key, reducers[key](state[key], action)),
-      Immutable(state)
-    )
+  // Validate reducers.
+  const validReducers = Object.keys(reducers).reduce((accum, key) => {
+    // A reducer must be a function.
+    if (typeof reducers[key] !== 'function') {
+      if (process.env.NODE_ENV !== 'production') {
+        const errorMessage = getNotSupportedTypeAsReducerError(reducers, key)
+        throw new Error(errorMessage)
+      }
+      return accum
+    }
+
+    return accum.set(key, reducers[key])
+  }, Immutable({}))
+
+  const validReducerKeys = Object.keys(validReducers)
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (validReducerKeys.length === 0) {
+      const errorMessage = getNoValidReducersError()
+      throw new Error(errorMessage)
+    }
+  }
+
+  let shapeAssertionError
+  try {
+    assertReducerShape(validReducers)
+  } catch (e) {
+    shapeAssertionError = e
+  }
+
+  return function combination(state = Immutable({}), action) {
+    if (shapeAssertionError) {
+      throw new Error(shapeAssertionError)
+    }
+
+    if (process.env.NODE_ENV !== 'production') {
+      const warningMessage = getPossibleUnexpectedStateShapeWarning(state, validReducers, action)
+      if (warningMessage) {
+        // eslint-disable-next-line no-console
+        console.tron ? console.tron.warn(warningMessage) : console.warn(warningMessage)
```

Comment:
is `console.tron` dependent on a library that people may not have? Would prefer straight up `console.log`

Reply:
I see that you are checking it exists first, but I don't like having an artefact from someone's development process in the codebase. Can we remove it?

Reply:
It is a logger application that can be used to get logs when developing for react-native. I'll take it out – no problem.

Reply:
@eadmundo this has been updated.